lib/dartdoc.dart

```diff
@@ -154,21 +155,27 @@ class DartDoc {
       });
     }
 
-    if (includes.isNotEmpty || excludes.isNotEmpty) {
-      print('generating docs for libraries ${libraries.join(', ')}\n');
-    }
-
-    Package package = new Package(libraries, packageMeta);
-
+    Package package;
     if (config != null && config.autoIncludeDependencies) {
-      final newLibraryElements =
-          _buildLibrariesWithAutoincludedDependencies(package);
-      Library.clearLibraryMap();
-      package = new Package(newLibraryElements, packageMeta);
+      package = Package.withAutoIncludedDependencies(libraries, packageMeta);
+      libraries = package.libraries.map((l) => l.element).toList();
+      // remove excluded libraries again, in case they are picked up through
+      // dependencies.
+      excludes.forEach((pattern) {
+        libraries.removeWhere((lib) {
+          return lib.name.startsWith(pattern) || lib.name == pattern;
+        });
+      });
     }
+    package = new Package(libraries, packageMeta);
+
+    print(
+        'generating docs for libraries ${package.libraries.map((Library l) => l.name).join(', ')}\n');
 
     // Go through docs of every model element in package to prebuild the macros index
-    package.allModelElements.forEach((m) => m.documentation);
+    // TODO(jcollins-g): move index building into a cached-on-demand generation
+    // like most other bits in [Package].
+    package.allCanonicalModelElements.forEach((m) => m.documentation);
```

Comment:
Eliminates repeats? Were we looking at documentation for elements that were not documented?